addon/mixins/route-query-manager.js

```diff
@@ -0,0 +1,11 @@
+import Ember from 'ember';
+import BaseQueryManager from 'ember-apollo-client/mixins/base-query-manager';
+
+const { Mixin } = Ember;
+
+export default Mixin.create(BaseQueryManager, {
+  resetController() {
+    this._super(...arguments);
+    this.get('apollo').unsubscribeAll();
```

Comment:
I'm facing a problem with this.

Let's say I'm on "/comments?status=approved" and then I navigate to "/posts":

1. the function `model` inside `app/routes/posts.js` runs `this.apollo.watchQuery` to fetch posts;
2. the query manager inside `app/routes/posts.js` keeps the active observable queries from this route;
3. the function `resetController` inside `app/routes/comments.js` runs `this.get('apollo').unsubscribeAll()`;
4. the query manager inside `app/routes/comments.js` unsubscribe all observable query.

That's fine and working as expected. Now let's say I'm on "/comments?status=approved" and then I navigate to "/comments?status=pending":

1. the function `model` inside `app/routes/comments.js` runs `this.apollo.watchQuery` to fetch pending comments;
2. the query manager inside `app/routes/comments.js` keeps the active observable queries from this route;
3. the function `resetController` inside `app/routes/comments.js` runs `this.get('apollo').unsubscribeAll()`;
4. the query manager inside `app/routes/comments.js` unsubscribe all observable query, including the one that was just pushed to the query manager.

The problem with this second scenario is noticed when I want to fetch more comments. The function `fetchMore` is called on the observable query that was previously unsubscribed and then Apollo tells me that the query does not exist.

I'm handle this using a copy of this mixin that only unsubscribe when we are exiting:

```javascript
resetController(_controller, isExiting, _transition) {
  this._super(...arguments);
  if (isExiting) {
    this.get('apollo').unsubscribeAll();
  }
}
```

Reply:
@viniciussbs ok, so I think I understand this scenario. When switching from `/comments?status=approved` to `/comments?status=pending`, you do in fact _want_ for the "approved comments" subscription to be ended. However, `resetController()`/`unsubscribeAll()` isn't called until __after__  the new model is loaded.

Is there maybe another route hook I should be utilizing to mark when a transition has begun? That would potentially let me flag all subscriptions which existed prior to the new model data, such that I would only clear things that existed before.

Reply:
@viniciussbs I just pushed up my progress thus far, including a new commit with a failing test case that I believe covers what you described. Can you please confirm before I write the fix?

My current plan is to utilize the `beforeModel()` route hook to tell the query manager to mark all pre-existing queries as stale, and then when `resetController()` gets called later, it should only cancel subscriptions that were marked as stale. Does that seem like the correct behavior and the best way to implement it?

Reply:
Sorry for the late response, @bgentry 

Yes, this is the scenario. I don't know a more appropriate hook for this case. Your plan looks like the ideal solution for me.